pumpkin-lib/src/propagators/arithmetic/maximum.rs

```diff
@@ -67,23 +69,120 @@ impl<ElementVar: IntegerVariable, Rhs: IntegerVariable> Propagator
         &self,
         mut context: PropagationContextMut,
     ) -> PropagationStatusCP {
+        // This is the constraint that is being propagated:
+        // max(a0, a1, ..., an-1) = rhs
+
+        // Propagate bounds on the variables based on the rhs:
+        // Rule 1.
+        // UB(a_i) <= UB(rhs)
+
+        // Propagate bounds on the rhs variable:
+        // Rule 2.
+        // LB(rhs) >= max{LB(a_i)}.
+        // Rule 3.
+        // max_ub = max{UB(a_i)}.
+        // if all UB(a_i) <= max_ub, then UB(rhs) <= max_ub.
+
         let rhs_ub = context.upper_bound(&self.rhs);
-        let mut max_ub = i32::MIN;
-        let mut max_lb = i32::MIN;
-        let mut lb_reason = vec![];
-        let mut ub_reason = vec![];
 
+        // Assume the variable at index zero is the minimum/maximum.
+        let mut max_ub = context.upper_bound(&self.array[0]);
+        let mut max_lb = context.lower_bound(&self.array[0]);
+        let mut lb_reason = predicate![self.array[0] >= max_lb];
         for var in self.array.iter() {
+            // Rule 1.
             context.set_upper_bound(var, rhs_ub, conjunction!([self.rhs <= rhs_ub]))?;
-            lb_reason.push(predicate![var >= context.lower_bound(var)]);
-            ub_reason.push(predicate![var <= context.upper_bound(var)]);
 
-            max_ub = i32::max(context.upper_bound(var), max_ub);
-            max_lb = i32::max(context.lower_bound(var), max_lb);
+            let var_lb = context.lower_bound(var);
+            let var_ub = context.upper_bound(var);
+
+            if var_lb > max_lb {
+                max_lb = var_lb;
+                lb_reason = predicate![var >= var_lb];
+            }
+
+            if var_ub > max_ub {
+                max_ub = var_ub;
+            }
+        }
+        // Rule 2.
+        context.set_lower_bound(&self.rhs, max_lb, PropositionalConjunction::from(lb_reason))?;
+
+        // Rule 3.
+        if rhs_ub > max_ub {
+            let ub_reason: PropositionalConjunction = self
+                .array
+                .iter()
+                .map(|var| predicate![var <= max_ub])
+                .collect();
+            context.set_upper_bound(&self.rhs, max_ub, ub_reason)?;
         }
 
-        context.set_upper_bound(&self.rhs, max_ub, PropositionalConjunction::new(ub_reason))?;
-        context.set_lower_bound(&self.rhs, max_lb, PropositionalConjunction::new(lb_reason))?;
+        // Rule 4.
+        // if LB(rhs) > UB(a_i) for all i, then conflict.
+        // This rule can be skipped because the previous rules establish UB(rhs) <= max{UB(a_i)}.
+
+        // Rule 5.
+        // If there is only one variable with UB(a_i) >= LB(rhs),
+        // then the bounds for rhs and that variable should be intersected.
+        let rhs_lb = context.lower_bound(&self.rhs);
+        let mut propagating_variable: Option<&ElementVar> = None;
+        let mut propagation_reason = PropositionalConjunction::default();
+        for var in self.array.iter() {
```

Comment:
Maybe? For now, my feeling is that it is acceptable since Maximum (my guess!) typically only has a small number of variables. If the Maximum turns out to show up in our tests we could look to improve the implementation.